plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -0,0 +1,282 @@
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.PriorityBlockingQueue;
+
+/**
+ * Service responsible for gathering, analyzing, storing and exporting
+ * top N queries with high latency data for search queries
+ *
+ * @opensearch.internal
+ */
+public class QueryInsightsService {
+    private static final Logger log = LogManager.getLogger(QueryInsightsService.class);
+
+    private static final TimeValue delay = TimeValue.ZERO;
+    /**
+     * The internal OpenSearch thread pool that execute async processing and exporting tasks
+     */
+    private final ThreadPool threadPool;
+
+    /**
+     * enable insight data collection
+     */
+    private final Map<MetricType, Boolean> enableCollect = new HashMap<>();
+
+    private int topNSize = QueryInsightsSettings.DEFAULT_TOP_N_SIZE;
+
+    private TimeValue windowSize = TimeValue.timeValueSeconds(QueryInsightsSettings.DEFAULT_WINDOW_SIZE);
+    /**
+     * The internal thread-safe store that holds the top n queries by latency query insight data
+     */
+    private final PriorityBlockingQueue<SearchQueryRecord> topQueriesByLatencyStore;
+    /**
+     * The internal thread-safe store that holds the top n queries by cpu query insight data
+     */
+    private final PriorityBlockingQueue<SearchQueryRecord> topQueriesByCpuStore;
+    /**
+     * The internal thread-safe store that holds the top n queries by jvm query insight data
+     */
+    private final PriorityBlockingQueue<SearchQueryRecord> topQueriesByJvmStore;
+    /**
+     * window start timestamp for each top queries collector
+     */
+    private final Map<MetricType, Long> topQueriesWindowStart = new HashMap<>();
+
+    /**
+     * Create the TopQueriesByLatencyService Object
+     *
+     * @param threadPool     The OpenSearch thread pool to run async tasks
+     */
+    @Inject
+    public QueryInsightsService(ThreadPool threadPool) {
+        this.topQueriesByLatencyStore = new PriorityBlockingQueue<>(
+            topNSize,
+            Comparator.comparingLong(a -> a.getMeasurement(MetricType.LATENCY).getValue().longValue())
+        );
+        this.topQueriesByCpuStore = new PriorityBlockingQueue<>(
+            topNSize,
+            Comparator.comparingDouble(a -> a.getMeasurement(MetricType.CPU).getValue().doubleValue())
+        );
+        this.topQueriesByJvmStore = new PriorityBlockingQueue<>(
+            topNSize,
+            Comparator.comparingDouble(a -> a.getMeasurement(MetricType.JVM).getValue().doubleValue())
+        );
+        this.threadPool = threadPool;
+    }
+
+    private PriorityBlockingQueue<SearchQueryRecord> getStoreByMetricType(MetricType type) {
+        if (type == MetricType.LATENCY) {
```

Comment:
nit: switch statement?

Reply:
Refacored the code to combine the priority queues into a map, in that way, I can remove this function.